archinstall/lib/installer.py

```diff
@@ -698,13 +709,17 @@ def _add_systemd_bootloader(self, root_partition: disk.PartitionModification):
 		# TODO: Ideally we would want to check if another config
 		# points towards the same disk and/or partition.
 		# And in which case we should do some clean up.
+		bootctl_options = [
+			f'--esp-path={efi_partition.mountpoint}' if efi_partition else '',
+			f'--boot-path={boot_partition.mountpoint}' if boot_partition else ''
+		]
 
 		# Install the boot loader
 		try:
-			SysCommand(f'/usr/bin/arch-chroot {self.target} bootctl --esp-path=/boot install')
+			SysCommand(f"/usr/bin/arch-chroot {self.target} bootctl {' '.join(bootctl_options)} install")
 		except SysCallError:
 			# Fallback, try creating the boot loader without touching the EFI variables
-			SysCommand(f'/usr/bin/arch-chroot {self.target} bootctl --no-variables --esp-path=/boot install')
+			SysCommand(f"/usr/bin/arch-chroot {self.target} bootctl --no-variables {' '.join(bootctl_options)} install")
 
 		# Ensure that the /boot/loader directory exists before we try to create files in it
 		loader_dir = self.target / 'boot/loader'
```

Comment:
```suggestion
		loader_dir = self.target / efi_partition / 'loader'
```

Do not forget to update this.

Reply:
The comment above it and the comment one the line before `entries_dir = loader_dir / 'entries'` should be updated too since they contain `/boot`. Maybe change `/boot` to `esp` in those comments.

Reply:
That's wise! Will do!

Reply:
hmm, isn't the path the same regardles of `efi_partition` or not? because `efi_partition` will be `/boot/EFI` in one case but `boot_partition` will be `/boot` in both. So I think keeping it makes sense. As the loader configuration will remain in `/boot/loader/entries/` in both instances?

Reply:
Take a look at how the ArchWiki refers to the path: https://wiki.archlinux.org/title/Systemd-boot#Loader_configuration

It uses <code><i>esp</i>/loader</code>, that is the source for my suggestion.

Reply:
https://wiki.archlinux.org/title/Systemd-boot#Installation_using_XBOOTLDR mentions:
> XBOOTLDR partition to `/mnt/boot`

And then to configure as normal, so `/boot/loader/entries` should naturally be on the `boot_partition` as XBOOTLDR partition will be found in `boot_partition`?

I might be confused tho, been bouncing between wiki entries and external sources on what the XBOOTLDR partition should or shouldn't be.

Reply:
At the top of the section you gave, emphasis mine:

> A separate [/boot partition](https://wiki.archlinux.org/title/Partitioning#/boot) of type "Linux extended boot" (XBOOTLDR) can be created to keep the **kernel** and **initramfs** _separate_ from the **ESP**.

Reply:
The ESP can be mounted at `/efi`, `/boot`, or `/boot/efi`. Using `/efi` is the most modern way and what I would suggest.

Reply:
I thought it was specifically for the EFI binary blobs, and it in turn checked `/boot/loader/loader.conf`. As the boot process worked for me I saw that as confirmation. So this is not the case?

Reply:
I can see why it would work on my test, as the loader *can* init via the EFI but finds no entries.

On the flip side, it could find entries but not load them?

Reply:
> - On disks with GPT (GUID Partition Table)
>
>   - The EFI System Partition (ESP for short) — a partition with a GPT type GUID of c12a7328-f81f-11d2-ba4b-00a0c93ec93b — may be used as one of two locations for boot loader menu entries.
>
>   - Optionally, an Extended Boot Loader Partition (XBOOTLDR partition for short) — a partition with GPT type GUID of bc13c2ff-59e6-4262-a352-b275fd6f7172 — may be used as the second of two locations for boot loader menu entries. This partition must be located on the same disk as the ESP.
    
https://uapi-group.org/specifications/specs/boot_loader_specification/#the-partitions

> It is recommended to mount $BOOT to /boot/, and the ESP to /efi/. If $BOOT and the ESP are the same, then either a bind mount or a symlink should be established making the partition available under both paths.

https://uapi-group.org/specifications/specs/boot_loader_specification/#mount-points

> A boot loader locates the XBOOTLDR partition and the ESP, then simply reads all the files /loader/entries/*.conf in them, and populates its boot menu (and handle gracefully if one of the two partitions is missing).

https://uapi-group.org/specifications/specs/boot_loader_specification/#locating-boot-entries

I would recommend doing a thorough read of this documentation and then read it once again. Any question you have on this topic should be answered in it.

Reply:
Sure, I can read more documentation.
But this expert:

> Optionally, an Extended Boot Loader Partition (XBOOTLDR partition for short) — a partition with GPT type GUID of bc13c2ff-59e6-4262-a352-b275fd6f7172 — may be used as the second of two locations for boot loader menu entries. This partition must be located on the same disk as the ESP.

In my mind confirms the current implementation. Where `XBOOTLDR` would be represented by the `boot_partition` in the code, *may* be used for boot loader menu entries.

The only check we might not do currently is to verify that the partition exists on the same disk as the ESP/EFI partition.

Or am I reading that incorrectly?

Reply:
It looks like when an XBOOTLDR is to be used, it is the correct location for the boot entries. I would still recommend against hard coding `/boot` at this location and in the comments.

Reply:
Reading it a third time, I am still convinced that the implementation is correct.
Namely this:

> If the ESP is separate from $BOOT it may also contain a /loader/entries/ directory, where the boot loader should look for boot entry snippets, as an additional source. The boot loader should enumerate both directories and present a merged list to the user. Note that this is done for compatibility only: while boot loaders should look in both places, **OSes should only add their files to $BOOT.**

ESP *(`efi_partition` in the code)* is separated from $BOOT *(`boot_partition`)*, seeing as both folders are enumerated, and that the recommendation is that the OS *(that would be us?)* should only add to $BOOT *(`boot_partition`)*, aka XBOOTLDR.

I still feel like you're questioning the implementation but the more I read it I start to believe that the implementation is sane?

Reply:
One reason why it perhaps would make sense to leave loader entries in `$BOOT/loader/entries` is that they're related to the installation, which `boot_partition` is *(it also happens to work out in our favor as the code can be simplified to deal with bot cases, `ESP` + `XBOOTLDR` and in the case of one united partition)*

Reply:
In my last comment I was agreeing with you. I also agree with your comments above, the implementation is sane.

This particular line of code we are commenting on should still have the `boot` portion of the string changed to a variable and not be hard coded. If an XBOOTLDR exists the variable should point to the mount point of it and if it does not is should point to the mount point of the ESP. Even if this ends up being `boot` for both or all cases let us not hard code it and acquire the path from a variable. For the comments that I referenced as needing to be updated, I recommend just dropping the `/boot/` prefix from the paths in them.

Reply:
> In my last comment I was agreeing with you.

Oh, I misunderstood it, my bad!
I think I put too much thought into this kinda small PR because it's been on the TODO since 2021 :S

> the variable should point to the mount point of it 

That's a good idea!